### ISSUE TITLE
Global amplitude is fixed, not learnt in multi experiments

### ISSUE THREAD

**Issue by tarepan** [ben-hayes/sinusoidal-gradient-descent#6]
## Summary
In multi-oscillator experiments, global amplitude a<sub>k</sub> (`global_amp`) is fixed, not learnt.  
This is caused by missing `.requires_grad_(True)`.  
Does this bug affect paper's results?  

## Bug
I try to run `evaluate_estimator -cn multi_fft_2`.  
In this setting, `use_global_amp` is True so global amplitude should be learnt.  

```yaml
# by `evaluate_estimator -cn multi_fft_2  --cfg job`
evaluation:
  _target_: sinusoidal_gradient_descent.eval.estimator.evaluation_loop
  use_global_amp: true
```

By inserting below debug codes in logging part, I check learning process of frequency and global amplitude. 
```python
341            if step % log_interval == 0:
342                print(f"Step {step}: {loss.item()}")
343                    if not use_real_sinusoid_baseline:
344                        if mode == "multi":
new                            a_k = torch.sigmoid(global_amp)
new                            print(a_k)
new                            point5 = torch.full(a_k.shape,  0.5, device=device)
new                            print(f"all 0.5: {torch.allclose(a_k, point5)}")
```

Frequency is learnt during iterative estimation, but global amplitude is not learnt, it is fixed to initial value.  
We got non-learnt `global_amp`.  

```
Step 0: 0.28140023350715637
tensor([[0.5000, 0.5000], ..., [0.5000, 0.5000]], device='cuda:0')
all 0.5: True
Freq error: 0.018339725211262703

Step 1000: 0.2812900245189667
tensor([[0.5000, 0.5000], ..., [0.5000, 0.5000]], device='cuda:0')
all 0.5: True
Freq error: 0.011968357488512993

...
Step 10000: 0.043224725872278214
tensor([[0.5000, 0.5000], ..., [0.5000, 0.5000]], device='cuda:0')
all 0.5: True
Freq error: 0.0067969998344779015
```

## Cause
The bug is caused by missing `.requires_grad_(True)`.  
https://github.com/ben-hayes/sinusoidal-gradient-descent/blob/cab2473307c710c28e11dd67f38d99f53607d0f1/sinusoidal_gradient_descent/eval/estimator.py#L291-L293

`z` is requires_grad==True, but `global_amp` is not.  

## Fix
I add `global_amp.requires_grad_(True)`, then `global_amp` starts to learn.  

```
Step 0: 0.28139975666999817
tensor([[0.5000, 0.5000], ..., [0.5000, 0.5000]], device='cuda:0')
all 0.5: False
Freq error: 0.018648464232683182

Step 1000: 0.28121888637542725
tensor([[0.5235, 0.5225], ..., [0.5246, 0.5241]], device='cuda:0')
all 0.5: False
Freq error: 0.012081814929842949
```

## Paper's results
This bug is caused by code, not config.  
Does this bug affect paper's results?  

<del>In my experiments about human voice analysis/synthesis, bigger |K| improve voice quality.  
This result could be inconsistent with paper's Fig.4 (good quality ~ low MSE).  
This inconsistency could be caused by this bug.</del>  
(this seems to be wrong. In Fig.4, bigger |K| make task itself more complex, so different from my experiment setup.)

**Comment by tarepan** [comment]
Hi @ben-hayes, sorry for the ping.

**Comment by ben-hayes** [comment]
Hey @tarepan, haven't forgotten about this, don't worry! We just had another deadline -- I plan to look into this further next week.

**Comment by ben-hayes** [comment]
Hi @tarepan 

Thanks again for submitting this issue. After looking into it, I can confirm that this did indeed have an effect on the results displayed in the paper. We are re-running experiments with the issue fixed (PR #7) and will include the correct results in the camera ready version for ICASSP. We will also upload a revision to arXiv. The good news is that, as you might expect, resolving this bug improves the estimator's performance, so the central findings of the paper remain very much intact.

If you're happy for me to do so, I would like to acknowledge your contribution in either the acknowledgements of the revised paper and on Twitter. Would you be willing to share your name and/or Twitter handle so I can do this?

**Comment by tarepan** [comment]
Better performance with bug fix, good news!  
I am looking forward to checking the result in updated paper.  

I feel honored to be acknowledged in the paper. My name is *Yuho Kobayashi*.  
And also, I feel happy to be shared in Twitter (only handle name is enough😉).  
My twitter name is *tarepan* and the account ID is `@tarepan_mhhow`.
